conftest.py

```diff
@@ -11,20 +11,72 @@
 ###############################################################################
 
 import sys
+import os
+import logging
+
+import pytest
+from typing import Mapping
+from hashlib import md5
 
 from spinalcordtoolbox.utils import sct_dir_local_path, sct_test_path
 sys.path.append(sct_dir_local_path('scripts'))
 
-import pytest
 import sct_download_data as downloader
 
 
-@pytest.fixture(scope='session', autouse=True)
-def download_data(request):
-    # This is a hack because the capsys fixture can't be used with
-    # session scope at the moment.
-    # https://github.com/pytest-dev/pytest/issues/2704#issuecomment-603387680
-    capmanager = request.config.pluginmanager.getplugin("capturemanager")
-    with capmanager.global_and_fixture_disabled():
-        print('\nDownloading sct testing data.')
-        downloader.main(['-d', 'sct_testing_data', '-o', sct_test_path()])
+logger = logging.getLogger(__name__)
+
+
+def pytest_sessionstart():
+    """ Download sct_testing_data prior to test collection. """
+    logger.info("Downloading sct test data")
+    downloader.main(['-d', 'sct_testing_data', '-o', sct_test_path()])
+
+
+@pytest.fixture(scope="session", autouse=True)
+def test_data_integrity(request):
+    files_checksums = dict()
+    for root, _, files in os.walk(sct_test_path()):
+        for f in files:
+            fname = os.path.join(root, f)
+            chksum = checksum(fname)
+            files_checksums[fname] = chksum
```

Comment:
From https://github.com/neuropoly/spinalcordtoolbox/pull/2903#issuecomment-696974432

> we've had problems in the past, where testing data were corrupted across time, and testing would not pass unless we would re-download the testing data

Since the issue described here is about maintaining the pure state of `sct_testing_data`, would it be useful to provide these checksums with the dataset itself instead?

If we did that, we could do something like  _"Only redownload `sct_testing_data` if the files in the existing `sct_testing_data` directory don't match the checksum."_ This could possibly allow us to not have to redownload `sct_testing_data` each time. (Although, I suppose it only takes a few seconds each time, so perhaps the savings are not worth the effort.)

Reply:
Yes definitely. I think there are 2 improvements to be made here:
1. Make sure that tests don't change the testing data (this PR).
2. Improve the downloader to support data caching, integrity checking, etc (another issue, to be discussed/designed).

For now, downloading the data on each run + checking integrity at the end ensures that the tests ran with the correct input and that they didn't cause side-effects on the input data.

Once the downloader is improved, I would expect `download_test_data()` to internally check if data is cached on disk and verify the integrity against the official checksums probably included in some kind of manifest, something that is not yet done to my knowledge.